### PR TITLE
arping: Fix exit code if receive more replies than sent

### PR DESCRIPTION
**Pull request by pevik**
ARP protocol, unlike ICMP protocol, has no way to link REQUEST and REPLY together (detect to which sender belongs the response). E.g. running more arping instances currently causes failure due receiving more replies than sent probes:

    # ./builddir/arping -c2 -I eth0 192.168.255.1 -w10 &
    # ./builddir/arping -c2 -I eth0 192.168.255.1 -w10 &
    ARPING 192.168.255.1 from 192.168.255.133 eth0
    ARPING 192.168.255.1 from 192.168.255.133 eth0
    Unicast reply from 192.168.255.1 [50:EB:F6:87:9D:D0]  1.722ms
    Unicast reply from 192.168.255.1 [50:EB:F6:87:9D:D0]  1.726ms
    Unicast reply from 192.168.255.1 [50:EB:F6:87:9D:D0]  1.910ms
    Unicast reply from 192.168.255.1 [50:EB:F6:87:9D:D0]  1.915ms
    Sent 1 probes (1 broadcast(s))
    Sent 1 probes (1 broadcast(s))
    Received 2 response(s)
    Received 2 response(s)
    [ ENTER ]
    [1]-  Exit 1                  ./builddir/arping -c2 -I eth0 192.168.255.1 -w10
    [2]+  Exit 1                  ./builddir/arping -c2 -I eth0 192.168.255.1 -w10

84ca65c (fix for 67e070d) introduced this regression.

Later e594ca5 introduced more precise timing - before arping sent 2 probes instead of 1 with -w1.

Then 854873b unified behavior with ping, i.e. using -w (deadline) *without* -c (count) exit 0 if at least one reply arrived (backwards incompatibility, also now incompatible with busybox). But that still kept problematic using -w with -c on multiple instances / replies.

Fixing the problem by adding a special condition.

Also, when at it, move all exit code evaluation into finish() (before it was in finish() but also event_loop()). This improves code introduced in 67e070d.

Fixes: 84ca65c ("arping: fix sent vs received packages return value")
Fixes: https://github.com/iputils/iputils/issues/538
Reported-by: Mingyang Liu <papillon@yeah.net>

### CODE REVIEW

File: arping.c [hunk 1]
```diff
@@ -317,11 +317,22 @@ static int finish(struct run_state *ctl)
...
+
 	if (ctl->unsolicited)
 		return 0;
-	return (!ctl->received);
+
+	if (ctl->timeout) {
+		if (ctl->count > 0)
+			return !(ctl->count <= ctl->received);
+		else
+			return !(ctl->received > 0);
```
Comment (cfconrad) [inline arping.c:332]: nitty, but you use `!(ctl->received > 0)` and `!ctl->received` I guess I would use only one of it everywhere...

and yes, what you mentioned here https://github.com/iputils/iputils/issues/538#issuecomment-2146244343 would help a lot :-)

Comment (pevik) [reply arping.c]: Thanks, correct! You're right, that `ctl->received > 0` can be shortened just to `ctl->received`, which makes the condition simpler (only `-c` together with `-w` counts how many replies we get):

```C
/* arping exit code evaluation */
if (ctl->dad)
	return !!ctl->received;

if (ctl->unsolicited)
	return 0;

if (ctl->timeout && ctl->count > 0)
	return !(ctl->count <= ctl->received);

return !ctl->received;
```

`ctl->count` and `ctl->interval` are by default `-1`, therefore they need to be checked against `0`.

> and yes, what you mentioned here https://github.com/iputils/iputils/issues/538#issuecomment-2146244343 would help a lot :-)

I'll guess I merge for now just this fix (to minimize changes) and send another PR which implements this + create exit code section for arping man page. There are many parts which could be improved, but it's a question how much of it should be done (as for all iputils tools: a complete rewrite would be the best to cleanup).

**Comment by michalsieron** [comment]
Hi, unfortunately this breaks `-f` flag, because in such case `ctl->count` being less than `ctl->received` is expected.
There should probably be a check for `ctl->quit_on_reply` somewhere.

**Comment by pevik** [comment]
@michalsieron @cfconrad @shuangye  Could you please retest?
@michalsieron Could you please send me your email address (either as a comment here or to my mail pvorel@suse.cz), so that I can add your Reported-by: and/or Reviewed-by: (ideally you would post these tag here yourself (e.g. https://github.com/iputils/iputils/pull/546#pullrequestreview-2098596101).

**Comment by michalsieron** [comment]
Yup. Seems to working just fine. Fixing both original issue with parallel arpinging and not regressing exit code when using `-f`. Tested after cherry-picking onto v20211215.

Reviewed-by: Michał Sieroń [michalwsieron@gmail.com](mailto:michalwsieron@gmail.com)

**Comment by cfconrad** [comment]
Nothing to add from my site, LGTM!

**Comment by xanoxes** [comment]
I have some questions about this merge. In the old realization, the 'ctl->brd_sent' can identify sendto failure in DAD mode. Due to the new realization, if sendto failed in sometime such like bond4 negotiation, we won't know from the return value.

**Comment by xanoxes** [comment]
https://github.com/iputils/iputils/issues/557

**Pull request merged**